### ISSUE TITLE
Some confusion about 'ctx_noparamgrad_and_eval' and optimizing noise

### ISSUE THREAD

**Issue by TianQi-777** [BorealisAI/advertorch#57]
Question 1:
In the `tutorial_train_mnist.py`, you designed the function [ctx_noparamgrad_and_eval()](https://github.com/BorealisAI/advertorch/blob/5daeeee98077afacf94728936a04cebde474259f/advertorch_examples/tutorial_train_mnist.py#L72), as follows:
```
with ctx_noparamgrad_and_eval(model):
    data = adversary.perturb(data, target)
```
Does it mean the same as follows:
```
model.eval()
with torch.no_grad():
    with torch.enable_grad():
        loss_attacker = ...
model.train()
```
I'm not sure if the above is correct. Is it true that `with torch.no_grad()` doesn't work in my code?  

I guess you write this function to prevent the calculation of redundant parameter gradients and the effect of `model.train()` when performing an attack, and at the same time, to ensure that the noise gradient can be calculated.

In the later evaluation, since there is no need to distinguish the parameter gradient and the noise gradient, they are unified to the following form: [Line 87](https://github.com/BorealisAI/advertorch/blob/5daeeee98077afacf94728936a04cebde474259f/advertorch_examples/tutorial_train_mnist.py#L87) and [Line 97](https://github.com/BorealisAI/advertorch/blob/5daeeee98077afacf94728936a04cebde474259f/advertorch_examples/tutorial_train_mnist.py#L97)
```
 model.eval()
with torch.no_grad():
```
I'm not sure if I understand correctly.

Question 2:
When attacking, your code is to optimize the noise separately [(as shown below)](https://github.com/BorealisAI/advertorch/blob/5daeeee98077afacf94728936a04cebde474259f/advertorch/attacks/iterative_projected_gradient.py#L65).
```
    delta.requires_grad_()
    for ii in range(nb_iter):
        outputs = predict(xvar + delta)
        loss = loss_fn(outputs, yvar)
```
I also found that there are some codes that adds the original image and noise first, and then performs overall optimization (as shown below). 
```
    x_adv = xvar + delta                           # Calculate the gradient of x_adv, not delta
    for ii in range(nb_iter):
        outputs = predict(x_adv)
        loss = loss_fn(outputs, yvar)
```
Is there a difference between these two methods?
looking forward to your reply : )

**Comment by gwding** [comment]
Re: question 1, your understanding is pretty much correct. It is basically `no_grad` on parameters.

Re: question 2, if you're not using pytorch optimizers (i.e. not treating `delta` as parameters), then using gradient from `xadv` and `delta` are equivalent, as they have the same gradient.

**Comment by TianQi-777** [comment]
Thanks for your quick reply.

Question 1: Is the following codes the same as `ctx_noparamgrad_and_eval()`?
```
loss_attacker = nn.CrossEntropyLoss()(xvar + delta , yvar, reduction = 'sum')
grad = torch.autograd.grad(loss_attacker, [delta])[0]
```
In this way, it is possible to differentiate only `delta`

Because you use `loss.backward()` in your code to obtain the gradient of `delta` through `delta.grad.data`, and `loss.backward()` will calculate the gradient of all parameters (including `model.parameters` and `delta`), you need to use `ctx_noparamgrad_and_eval()` to explicitly decouple the gradient of the `model parameters` and the gradient of `delta`. Is using `torch.autograd.grad` an alternative method?

Question 2:
I do use pytorch optimizers, I'm not sure if there will be a difference in optimizing the `delta` and optimizing (`xvar + delta`) in the pytorch framework.

**Comment by gwding** [comment]
Re: Q1, the method you proposed should achieve similar purpose, but it is not exactly the same as `ctx_noparamgrad_and_eval()`. To give a simple example, let's say `loss = x**2+y**2`
and we want to take gradient of `x` but not `y`.

The method you propose is to turn on `requires_grad` for both `x` and `y`, and only calculate gradient for `y`. This way, you save some computation in backpropagation, but in the forward pass, you still need to save the computational graph of `y**2`.

When using `ctx_noparamgrad_and_eval()`, you just need to save the value of `y**2`, not it's graph.

Re: Q2, mathematically, those 2 are exactly the same. The difference is only at the implementation level, which I also think there shouldn't be a difference.

**Comment by TianQi-777** [comment]
Ok, I understand. Thank you for your careful reply :)
